### ISSUE TITLE
[STRI 1] Possible to run multiple contracts for free

### ISSUE THREAD

**Issue by C0x41lch0x41** [stellar/stellar-turrets#3]
### What version are you using?
0.0.0

### Scenario
A user with enough balance to make a request for just one run could
potentially create thousands of concurrent requests for runs that will be accepted, up
until the first one that finishes and returns subtracting the cost from the user
balance, though by the time that happens there will be many contracts running for
free.

### Why
The user balance is decreased after the run finishes. This creates a window to request many contract runs with a balance for just one request.

### Recommendation
Subtracting the maximum possible cost before running a contract and refunding the remainder after. To lock this further we could add a concurrency lock to the users balances during the time the balance is updated in the store.

**Comment by silence48** [comment]
This issue is resolved in commit (https://github.com/stellar/stellar-turrets/pull/2/commits/47339daf33bee887ee95233f6d0c6d63f7c72785) 

The commit actually encompasses both STRI 1 and STRI 2.  By utilizing the lambda runtime timeout  and the turret run divisor
https://github.com/stellar/stellar-turrets/blob/47339daf33bee887ee95233f6d0c6d63f7c72785/wrangler/src/txFunctions/run.js#L17-L18

Using these datapoints, we are able to calculate a reserve fee which is the maximum cost to run the contract.
https://github.com/stellar/stellar-turrets/blob/47339daf33bee887ee95233f6d0c6d63f7c72785/wrangler/src/txFunctions/run.js#L57-L59

which is subtracted from the user balance before runtime https://github.com/stellar/stellar-turrets/blob/47339daf33bee887ee95233f6d0c6d63f7c72785/wrangler/src/txFunctions/run.js#L71-L79

then the difference is credited back after runtime https://github.com/stellar/stellar-turrets/blob/47339daf33bee887ee95233f6d0c6d63f7c72785/wrangler/src/txFunctions/run.js#L133-L141

The concurrency issue is handled in that the reserve fee is subtracted right before the runner is called, and there is a check to ensure that the users available fee balance is more than the reserve fee 
https://github.com/stellar/stellar-turrets/blob/47339daf33bee887ee95233f6d0c6d63f7c72785/wrangler/src/txFunctions/run.js#L64-L69

This ensures that there is always a fee balance available before starting a runner.

**Comment by C0x41lch0x41** [comment]
This is patched here: https://github.com/stellar/stellar-turrets/pull/11 and merged into master here: https://github.com/stellar/stellar-turrets/commit/cf6de0e32adfd2e6f1ff73a8eaa8e8cce9de00de
